File: app/src/main/java/com/settlex/android/ui/dashboard/viewmodel/PromoBannerViewModel.java
package com.settlex.android.ui.dashboard.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.settlex.android.data.repository.PromoBannerRepository;
import com.settlex.android.ui.dashboard.model.PromoBannerUiModel;
import com.settlex.android.util.event.UiState;

import java.util.List;

import dagger.hilt.android.lifecycle.HiltViewModel;
import jakarta.inject.Inject;

@HiltViewModel
public class PromoBannerViewModel extends ViewModel {
    private final PromoBannerRepository promoBannerRepo;
    private final MutableLiveData<UiState<List<PromoBannerUiModel>>> promoBannersLiveData = new MutableLiveData<>();

    @Inject
    public PromoBannerViewModel(PromoBannerRepository promoBannerRepo) {
        this.promoBannerRepo = promoBannerRepo;
        fetchPromoBanners();
    }

    // Getters
    public LiveData<UiState<List<PromoBannerUiModel>>> getPromoBanners() {
        return promoBannersLiveData;
    }

    private void fetchPromoBanners() {
        promoBannersLiveData.postValue(UiState.loading());
        promoBannerRepo.fetchPromotionalBanners(bannerUiModel -> promoBannersLiveData.postValue(UiState.success(bannerUiModel)));
    }
}

File: app/src/main/java/com/settlex/android/ui/dashboard/viewmodel/PromoBannerViewModel.kt
package com.settlex.android.ui.dashboard.viewmodel

import androidx.lifecycle.ViewModel
import com.settlex.android.data.repository.BannerRepository
import com.settlex.android.ui.common.state.UiState
import com.settlex.android.ui.dashboard.model.BannerUiModel
import dagger.hilt.android.lifecycle.HiltViewModel
import jakarta.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

@HiltViewModel
class PromoBannerViewModel @Inject constructor(private val bannerRepository: BannerRepository) :
    ViewModel() {
    private val _bannerState =
        MutableStateFlow<UiState<MutableList<BannerUiModel>>>(UiState.Loading)
    val bannerState = _bannerState.asStateFlow()

    init {
        fetchPromoBanners()
    }

    private fun fetchPromoBanners() {
        bannerRepository.fetchPromotionalBanners { bannerUiModel: MutableList<BannerUiModel> ->
            _bannerState.value = UiState.Success(bannerUiModel)
        }
    }
}
